### ISSUE TITLE
Export Stalls

### ISSUE THREAD

**Issue by spacehamster** [DaZombieKiller/TypeTreeDumper#2]
I've started to implement exporting here https://github.com/spacehamster/TypeTreeDumper/tree/working, and issues with code that I would have expected to cause a hard crash cause the process to stall. Closing the terminal leaves a zombie unity process alive. There are no stack traces logged in Editor.log either. I wonder if there are good ways to debug issues like that?

**Comment by DaZombieKiller** [comment]
I took a quick look through the code, and I saw an issue that could be causing stalls: `Produce` was changed to take `in byte` instead of `in RuntimeTypeInfo`. `in` can cause a copy to be made, which would truncate the data sent to the function to the size of a `byte`. It should use `ref` instead of `in`.

I also noticed that the managed delegate for `GenerateTypeTree` is wrong. The order for that one should be `object, tree, flags` instead of `object, flags, tree`. Unless you're testing with a version of Unity that lacks `TypeTreeCache` I don't think that's the cause of the hang though.

The zombie process could probably be dealt with by terminating the Unity process from the main console app when it closes. I'm not sure about how other issues could be debugged though.

**Comment by spacehamster** [comment]
Thanks for the help with Produce and GenerateTypeTree. Debugging with logging to console had narrowed it down to something to do with object produce, I wasn't aware I had screwed up GenerateTypeTree too.

The question was mostly about debugging, I'm not very experienced with C# interop, so I expect to make a bunch of dumb mistakes like that, and debugging is a bit harder without the stacktraces unity was giving when it crashed. I'll look into it a bit more and see if I can come up with a good solution.

**Comment by DaZombieKiller** [comment]
I did some cursory searching, and apparently adding `-logFile` (with no file name listed afterwards) will cause Unity to route all logging to `stdout`, so you could potentially retrieve any stack traces in the main console program by reading from it.

**Comment by spacehamster** [comment]
Ah, I had assumed that it was automatically writing logs to `C:\Users\username\AppData\Local\Unity\Editor\Editor.log`, but it seems like you need to manually set a log path. I had been reading stale log files.

After fixing the log path and the Object::Produce signature, the log shows:
```
Cannot create on non-main thread without kCreateObjectFromNonMainThread

Assertion failed on expression: 'CurrentThread::IsMainThread()
```
and stalls while calling destroy immediate. It seems it doesn't like creating and destroying objects from outside the main thread.

Also passing ObjectCreationMode.FromNonMainThread to Object::Produce changes the log to this.
```
Cannot create on non-main thread without an instanceID

Assertion failed on expression: 'CurrentThread::IsMainThread()'
```

**Comment by DaZombieKiller** [comment]
Yeah, you need to create an instance ID yourself if you want to create an object outside of the main thread. I *think* there's a function for that somewhere in the code, I just don't know the signature.

For destroying objects outside of the main thread, we might be able to make use of the dummy project. Unity has an `-executeMethod` parameter that can be used to call a static method after loading the project, which could connect to the injected TypeTreeDumper.Client assembly somehow and provide an API for queuing work on the main thread.

**Comment by spacehamster** [comment]
I was going to suggest trying to use -executeMethod to trigger the main thread to somehow call into `TypeTreeDumper.Client` to get it running on the main thread, but if EasyHook can hook into the update loop, that would be even better.

**Comment by DaZombieKiller** [comment]
The `AfterEverythingLoaded` callback might actually be running on the main thread too, so it's worth experimenting with running more code in that.

**Comment by DaZombieKiller** [comment]
Something interesting to note is that using `SymbolResolver.Resolve`/`SymbolResolver.ResolveFunction` within that function can cause a hang. I think the cause is DIA, because if I resolve that symbol in the `Run` function first, there is no hang (`DiaSymbolResolver` caches previously resolved symbols, so DIA wouldn't run the second time). I have no idea why that would happen though.

**Comment by spacehamster** [comment]
Hmm, I wonder if dbghelp.dll has the same issue. The thread issue probably isn't big enough problem to switch over even if it doesn't.

**Comment by DaZombieKiller** [comment]
I just pushed two commits that fix cross-thread usage of `DiaSymbolResolver`, it was actually much simpler than the solution I just proposed above.

**Comment by spacehamster** [comment]
> The AfterEverythingLoaded callback might actually be running on the main thread too, so it's worth experimenting with running more code in that.
Edit: I tried printing the result of CurrentThread::IsMainThread in that callback and got True.

i tried calling `?IsMainThread@CurrentThread@@YA_NXZ` and it returns true when called inside the AfterEverythingLoaded callback *and* when called from `EntryPoint.Run`, there seems to be something wrong with that call.

Moving the dumping logic to the AfterEverythingLoaded callback fixes the stall.

The log changes from
```
Cannot create on non-main thread without an instanceID

Assertion failed on expression: 'CurrentThread::IsMainThread()'```
```
to a 295 lines of
 ```Cannot create on non-main thread without an instanceID```

The output is not valid, but i'm still looking into that

**Comment by DaZombieKiller** [comment]
> I tried calling `?IsMainThread@CurrentThread@@YA_NXZ` and it returns true when called inside the AfterEverythingLoaded callback and when called from `EntryPoint.Run`, there seems to be something wrong with that call.

This is caused by an incorrect P/Invoke signature. The default behaviour for `bool` is to marshal a WinAPI `BOOL` type, instead of a C++ `bool`. To fix this, use `[return: MarshalAs(UnmanagedType.U1)]`:

```CSharp
[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
[return: MarshalAs(UnmanagedType.U1)]
delegate bool IsMainThreadDelegate();
```

**Comment by spacehamster** [comment]
Regarding the export stall with Unity 5.6.7, I created a c++ dumper to help with debugging https://github.com/spacehamster/NativeTypeTreeDumper.
When using TypeTreeDumper, the editor log doesn't have a stacktrace, while the c++ dumper does.

TypeTreeDumper [Editor.log](https://github.com/DaZombieKiller/TypeTreeDumper/files/5213520/Editor.log)
C++ Dumper [Editor.log](https://github.com/DaZombieKiller/TypeTreeDumper/files/5213521/Editor.log)

The stacktrace looks like this
```
0x000000014104B685 (Unity) MemoryProfiler::RegisterRootAllocation
0x0000000140784077 (Unity) assign_allocation_root
0x000000014013E511 (Unity) BaseObjectInternal::NewObject<AssetMetaData>
0x00000001401450B2 (Unity) ProduceHelper<AssetMetaData,0>::Produce
0x0000000140927067 (Unity) Object::Produce
```

and some testing shows that setting MemLabelId.Identifier to 0x32 stops the stall in TypeTreeDumper and stops the crash in the c++ dumper. An identifier of 0x32 is equivalent to `?kMemBaseObject@@3UMemLabelId@@A` in 5.6.7, but I have not checked more recent versions.

**Comment by DaZombieKiller** [comment]
> 
> 
> When using TypeTreeDumper, the editor log doesn't have a stacktrace, while the c++ dumper does.

I wonder if there's a way for us to still have a native stack trace in managed code. I can't imagine it's easy though, and probably requires a bunch of managed<->unmanaged hopping around.

**Comment by spacehamster** [comment]
I'm looking at how to handle STL strings for versions 5.4 and lower. basic_string::c_str has a slightly different signature in each version.
5.4:
```
PublicSymbol: [0000EAB0][0001:0000DAB0] 
?c_str@?$basic_string@DU?$char_traits@D@std@@V?$stl_allocator@D$0EC@$0BA@@@@std@@QEBAPEBDXZ
public: char const * __cdecl std::basic_string<char,struct std::char_traits<char>,class stl_allocator<char,66,16> >::c_str(void)const
```
5.3:
```
PublicSymbol: [00D11560][0001:00D10560] 
?c_str@?$basic_string@DU?$char_traits@D@std@@V?$stl_allocator@D$0EB@$0BA@@@@std@@QEBAPEBDXZ
public: char const * __cdecl std::basic_string<char,struct std::char_traits<char>,class stl_allocator<char,65,16> >::c_str(void)const
```
5.2:
```
PublicSymbol: [0034B010][0001:0034A010] 
?c_str@?$basic_string@DU?$char_traits@D@std@@V?$stl_allocator@D$0CP@$0BA@@@@std@@QEBAPEBDXZ
public: char const * __cdecl std::basic_string<char,struct std::char_traits<char>,class stl_allocator<char,47,16> >::c_str(void)const

PublicSymbol: [0034B010][0001:0034A010]
?c_str@?$basic_string@DU?$char_traits@D@std@@V?$stl_allocator@D$0DL@$0BA@@@@std@@QEBAPEBDXZ
public: char const * __cdecl std::basic_string<char,struct std::char_traits<char>,class stl_allocator<char,59,16> >::c_str(void)const 
```
Etcetera, so I think symbol resolver needs to be extended to support finding symbols that match a prefix (starting with `?c_str@?$basic_string`).

**Comment by DaZombieKiller** [comment]
```cs
public partial class SymbolResolver
{
    public abstract string[] FindSymbolsWithPrefix(string prefix);
}
```
I think an API like this would be fine. There can also be a few helper methods such as:
```cs
public partial class SymbolResolver
{
    public T* ResolveFirstWithPrefix<T>(string prefix) where T : unmanaged;
    public T  ResolveFirstFunctionWithPrefix<T>(string prefix) where T : Delegate;
}

**Comment by DaZombieKiller** [comment]
[Just pushed support for this on master](https://github.com/DaZombieKiller/TypeTreeDumper/commit/ffb7b00d58fdc192dbcc61dcbe9451804733f061). Since DIA supports Regex, I changed the API design a little bit to reflect that.

**Comment by spacehamster** [comment]
This line doesn't print out the exception
https://github.com/DaZombieKiller/TypeTreeDumper/blob/fe16fd931a4d78331fa8a5de180cd9ae9e045592/TypeTreeDumper.Client/EntryPoint.cs#L80

but if you change it to `Console.Error.WriteLine(ex.ToString());`, it does. i'm not sure why that would be the case

**Comment by DaZombieKiller** [comment]
That's strange, I wonder if it's because it's being set to `IpcInterface.Error` which is tied to the server, and thus it can only take certain types. We might have to make a wrapper `TextWriter` to work around that. If what I'm thinking is correct, then `Console.WriteLine(ex);` should work, but `Console.Out.WriteLine(ex);` shouldn't.

**Comment by DaZombieKiller** [comment]
Hm, I still see exceptions printed out even with `Console.Error.WriteLine(ex);`. Do you have an example of a situation where it fails so I can experiment with it?

**Comment by spacehamster** [comment]
I just add `throw new UnresolvedSymbolException("Missing Symbol Name Here"); to the top of 
```csharp
void ExecuteDumper()
{
    throw new UnresolvedSymbolException("Missing Symbol Name Here");
    var GetUnityVersion   = resolver.ResolveFunction<GetUnityVersionDelegate>("?GameEngineVersion@PlatformWrapper@UnityEngine@@SAPEBDXZ");
    var ParseUnityVersion = resolver.ResolveFunction<UnityVersionDelegate>("??0UnityVersion@@QEAA@PEBD@Z");
    ParseUnityVersion(out UnityVersion version, Marshal.PtrToStringAnsi(GetUnityVersion()));
    Dumper.Execute(new UnityEngine(version, resolver), server.OutputDirectory);
}
```
It also appears that if I change UnresolvedSymbolException to System.Exception, it prints like normal.

**Comment by DaZombieKiller** [comment]
I can confirm that a wrapper TextWriter fixes the issue. Fixed in https://github.com/DaZombieKiller/TypeTreeDumper/commit/81d2e56c4f1be16a041fede1c55e9927ccf32a46.

**Comment by spacehamster** [comment]
Do you know why it effects UnresolvedSymbolException but not System.Exception?

**Comment by DaZombieKiller** [comment]
That's because `UnresolvedSymbolException` is a custom exception that doesn't implement serialization, which would be necessary for it to travel between processes.

**Comment by spacehamster** [comment]
Remote hooking Unity 4.7 doesn't work, a blank console pops up and nothing happens. I'm guessing it is because it is a 32 bit app? I believe 5.0 onwards are 64 bit.

**Comment by DaZombieKiller** [comment]
That's probably why, yeah. It would probably work if you enable `Prefer 32 Bit` for a build, and you'll need to register the 32-bit msdia140.dll as well.

**Comment by DaZombieKiller** [comment]
The [most recent commit on master](https://github.com/DaZombieKiller/TypeTreeDumper/commit/9d4bd6be9fce655b706d14d17c6c0339cbe24675) should now work when the dumper is compiled for `x86` with no further changes.

**Comment by spacehamster** [comment]
WIth Unity 4.7, ClassIDToRTTI always returns null. I've not been able to figure out why.

**Comment by DaZombieKiller** [comment]
> WIth Unity 4.7, ClassIDToRTTI always returns null. I've not been able to figure out why.

It might require some investigation in Ghidra.

> Also, side note, AfterEverythingLoaded was changed to a __thiscall in 4.7.

Thanks for the heads up, I haven't verified most of the calling conventions since it only applies to x86 and not x64.

**Comment by spacehamster** [comment]
The function seems really simple
```c
/* public: static struct Object::RTTI * __cdecl Object::ClassIDToRTTI(int) */

RTTI * __cdecl ClassIDToRTTI(int param_1)

{
  _Tree<class_std::_Tmap_traits<int,struct_Object::RTTI,struct_std::less<int>,class_stl_allocator<struct_std::pair<int_const_,struct_Object::RTTI>,1,4>,0>_>
  *p_Var1;
    
  _Tree_iterator<std::_Tree_val<std::_Tmap_traits<int,Object::RTTI,std::less<int>,stl_allocator<std::pair<intconst,Object::RTTI>,1,4>,0>>>
  i;
  
  p_Var1 = gRTTI;
  find(gRTTI,(int *)&i);
  if (i == *(
             _Tree_iterator<std::_Tree_val<std::_Tmap_traits<int,Object::RTTI,std::less<int>,stl_allocator<std::pair<intconst,Object::RTTI>,1,4>,0>>>
             *)(p_Var1 + 4)) {
    return (RTTI *)0x0;
  }
  return (RTTI *)((int)i + 0x10);
}
```
(BTW i renamed gRTTI, it doesn't have a symbol associated with it)

My first guess is that gRTTI was not being initialized, so I tried calling RegisterAllClasses and InitializeAllClasses, but that does help (it warns that it can't register classes multiple times. Changing the entry point to InitializeEngineNoGraphics also didn't help

**Comment by DaZombieKiller** [comment]
I'm assuming the secret lies in whatever that `find` function is, since the `null` return seems dependent on whatever it sets `i` to.

**Comment by spacehamster** [comment]
I think I was having a brain fart with the `ClassIDToRTTI` issue and just misreading the logs, it seems to be working now.

Unity editor 4.0 - 4.3 crash on launch because of a bug in the licensing server producing invalid licenses [see here for details](https://forum.unity.com/threads/recent-changes-to-unity-license-server-output-broke-unity-4-3.819297/). This can be worked around with a detour on one of the licensing manager methods, but I wonder if there is a way to get unity to fix their licensing server.

EasyHook is unable to hook AfterEverythingLoaded in versions 3.4 and 3.5.
```
System.NotSupportedException: STATUS_NOT_SUPPORTED: Hooking near conditional jumps is not supported. (Code: 0)
   at EasyHook.LocalHook.Create(IntPtr InTargetProc, Delegate InNewProc, Object InCallback)
   at TypeTreeDumper.EntryPoint.Run(IContext context, String channelName)
```
A bunch of methods are not supported by EasyHook because of jumps near the start of the function.
Attempting to hook InitializeEngineGraphics or CreateMainWindowForBatchMode (which don't throw exceptions when creating hooks) cause the editor to crash. Ghidra also has a lot of trouble decomping versions 3.4 and 3.5. it seems they are doing some weird stuff in the native code.

A workaround is to do this, but its kinda clunky:

Program.cs
```csharp
var scriptLoader = "";
var versionInfo = FileVersionInfo.GetVersionInfo(args[0]);
string version = versionInfo.FileVersion;
if (version.StartsWith("3."))
{
	scriptLoader = " -executeMethod Loader.Load";
}
RemoteHooking.IpcCreateServer(ref channel, WellKnownObjectMode.Singleton, server);
RemoteHooking.CreateAndInject(args[0],
	$"-nographics -batchmode -{command} \"{project}\" -logFile \"{logPath}\"",
	InProcessCreationFlags: 0,
	InjectionOptions.DoNotRequireStrongName,
	typeof(EntryPoint).Assembly.Location,
	typeof(EntryPoint).Assembly.Location,
	out int processID,
	channel
);
```

EntryPoint
```csharp
[SuppressMessage("Style", "IDE0060", Justification = "Required by EasyHook")]
public void Run(RemoteHooking.IContext context, string channelName)
{
	try
	{
		//PublicSymbol: [00C149A0][0001:00C139A0] ?AfterEverythingLoaded@Application@@QAEXXZ(public: void __thiscall Application::AfterEverythingLoaded(void))
		try
		{
			var patten = new Regex(Regex.Escape("?AfterEverythingLoaded@Application@") + "*");
			var address = resolver.ResolveFirstMatching(patten);
			var hook = LocalHook.Create(address, new AfterEverythingLoadedDelegate(AfterEverythingLoaded), null);
			hook.ThreadACL.SetExclusiveACL(Array.Empty<int>());
		} catch(NotSupportedException)
		{
			Console.WriteLine("Could not hook AfterEverythingLoaded, using Script Loader");
			InitLoader();
		}
		// ... Snip
	}
}

delegate void ExecuteDumperDelegate();

void InitLoader()
{
	var ptr = Marshal.GetFunctionPointerForDelegate(new ExecuteDumperDelegate(ExecuteDumper));
        //TODO: how to get project path from entry point?
	File.WriteAllText(@"C:\Code\TypeTreeDumper\TypeTreeDumper\bin\Debug\net471\DummyProject\Assets\Loader.cs",
		@"
using System;
using System.IO;
using System.Runtime.InteropServices;
using UnityEditor;


public class Loader
{
    delegate void ExecuteDumperDelegate();
    public static void Load()
    {
        File.WriteAllText(""Loader.txt"", ""Hello world!"");
        var ptr = new IntPtr(ADDRESS);
        var del = Marshal.GetDelegateForFunctionPointer(ptr, typeof(ExecuteDumperDelegate));
        del.DynamicInvoke();
        EditorApplication.Exit(0);
    }
}".Replace("ADDRESS", ptr.ToInt64().ToString()));
}
```

**Comment by DaZombieKiller** [comment]
> A workaround is to do this, but its kinda clunky:

An alternative could be to include EasyHook and the `TypeTreeDumper.Client.dll` file in the Unity project (so it has access to the `IpcInterface` class). Then we supply the channel name to Unity via the command line, which we could use to retrieve the `IpcInterface` instance, which we'd store a function pointer in.

```CSharp
using System;
using System.Runtime.InteropServices;
using TypeTreeDumper;
using EasyHook;

public static class LegacyEntryPoint
{
    static IpcInterface server;

    public static void Execute()
    {
        var args    = Environment.GetCommandLineArgs();
        var channel = string.Empty;

        for (int i = 0; i < args.Length; i++)
        {
            if (args[i] == "-hook-channel")
            {
                channel = args[i + 1];
                break;
            }
        }

        server = RemoteHooking.IpcConnectClient<IpcInterface>(channel);
        // Now use the pointer stored on IpcInterface to execute the dumper
    }
}
```

**Comment by spacehamster** [comment]
I'm worried that trying to use TypeTreeDumper.Client.dll and IpcInterface from within the unity mono runtime will cause a bunch of compatibility issues. It'd probably have to just be `LegacyEntryPoint` and `IpcInterface ` in alone in a .NET 2.0 assembly. 

The reason i think its clunky (besides the obvious raw pointer function calls) is that you need to do feature detection twice, first in TypeTreeDumper to set the commandline flags `-executeMethod Loader.Load` and again in TypeTreeDumper.Client to initialise the loading script. There does not seem to be an [InitializeOnLoad] attribute for 3.5. If we could modify the commandline flags from within TypeTreeDumper.Client, we could remove the version check from TypeTreeDumper. Another option is to add the entry method to the IpcInterface and remove the try catch block from TypeTreeDumper.Client.

**Comment by DaZombieKiller** [comment]
> I'm worried that trying to use EasyHook and TypeTreeDumper.Client.dll from within the unity mono runtime will cause a bunch of compatibility issues.

As long as the .NET 2.0 version is used, EasyHook should work fine. I'm not sure about the dumper assembly though, we might need to make another .NET 2.0 dll to contain the `IpcInterface` class and any other types that need to be accessed by the legacy entry point.

> The reason I think its clunky (besides the obvious raw pointer function calls) is that you need to do feature detection twice

Yeah, this is definitely something we should solve. Maybe we could just unconditionally pass `-executeMethod` and then decide what to do based on information retrieved from `IpcInterface`?

**Comment by spacehamster** [comment]
I tried to write a Legacy Entry Point POC [here](https://github.com/spacehamster/TypeTreeDumper/tree/loader) but I get a binary serialization error when trying to access the IPC. 

```
System.Runtime.Serialization.SerializationException: Invalid binary format

Server stack trace: 
  at System.Runtime.Serialization.Formatters.Binary.ObjectReader.ReadRefTypeObjectInstance (System.IO.BinaryReader reader, System.Int64& objectId, System.Object& value, System.Runtime.Serialization.SerializationInfo& info) [0x00000] in <filename unknown>:0 
  at System.Runtime.Serialization.Formatters.Binary.ObjectReader.ReadObject (BinaryElement element, System.IO.BinaryReader reader, System.Int64& objectId, System.Object& value, System.Runtime.Serialization.SerializationInfo& info) [0x00000] in <filename unknown>:0 
  at System.Runtime.Serialization.Formatters.Binary.ObjectReader.ReadNextObject (BinaryElement element, System.IO.BinaryReader reader) [0x00000] in <filename unknown>:0 
  at System.Runtime.Serialization.Formatters.Binary.ObjectReader.ReadObjectGraph (BinaryElement elem, System.IO.BinaryReader reader, Boolean readHeaders, System.Object& result, System.Runtime.Remoting.Messaging.Header[]& headers) [0x00000] in <filename unknown>:0 
  at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.NoCheckDeserialize (System.IO.Stream serializationStream, System.Runtime.Remoting.Messaging.HeaderHandler handler) [0x00000] in <filename unknown>:0 
  at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Deserialize (System.IO.Stream serializationStream) [0x00000] in <filename unknown>:0 
  at System.Runtime.Remoting.Channels.Ipc.Win32.IpcTransport.Read (ITransportHeaders& headers, System.IO.Stream& responseStream) [0x00000] in <filename unknown>:0 
  at System.Runtime.Remoting.Channels.Ipc.Win32.IpcClientChannelSink.ProcessMessage (IMessage msg, ITransportHeaders requestHeaders, System.IO.Stream requestStream, ITransportHeaders& responseHeaders, System.IO.Stream& responseStream) [0x00000] in <filename unknown>:0 
  at System.Runtime.Remoting.Channels.BinaryClientFormatterSink.SyncProcessMessage (IMessage msg) [0x00000] in <filename unknown>:0 

Exception rethrown at [0]: 

  at System.Runtime.Serialization.Formatters.Binary.ObjectReader.ReadRefTypeObjectInstance (System.IO.BinaryReader reader, System.Int64& objectId, System.Object& value, System.Runtime.Serialization.SerializationInfo& info) [0x00000] in <filename unknown>:0 
  at System.Runtime.Serialization.Formatters.Binary.ObjectReader.ReadObject (BinaryElement element, System.IO.BinaryReader reader, System.Int64& objectId, System.Object& value, System.Runtime.Serialization.SerializationInfo& info) [0x00000] in <filename unknown>:0 
  at System.Runtime.Serialization.Formatters.Binary.ObjectReader.ReadNextObject (BinaryElement element, System.IO.BinaryReader reader) [0x00000] in <filename unknown>:0 
  at System.Runtime.Serialization.Formatters.Binary.ObjectReader.ReadObjectGraph (BinaryElement elem, System.IO.BinaryReader reader, Boolean readHeaders, System.Object& result, System.Runtime.Remoting.Messaging.Header[]& headers) [0x00000] in <filename unknown>:0 
  at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.NoCheckDeserialize (System.IO.Stream serializationStream, System.Runtime.Remoting.Messaging.HeaderHandler handler) [0x00000] in <filename unknown>:0 
  at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Deserialize (System.IO.Stream serializationStream) [0x00000] in <filename unknown>:0 
  at System.Runtime.Remoting.Channels.Ipc.Win32.IpcTransport.Read (ITransportHeaders& headers, System.IO.Stream& responseStream) [0x00000] in <filename unknown>:0 
  at System.Runtime.Remoting.Channels.Ipc.Win32.IpcClientChannelSink.ProcessMessage (IMessage msg, ITransportHeaders requestHeaders, System.IO.Stream requestStream, ITransportHeaders& responseHeaders, System.IO.Stream& responseStream) [0x00000] in <filename unknown>:0 
  at System.Runtime.Remoting.Channels.BinaryClientFormatterSink.SyncProcessMessage (IMessage msg) [0x00000] in <filename unknown>:0 ```

Also copying the legacy entry point into the unity project seems unnecessarily complicated. It could probably be done as a build event, but still feels less that celan.

**Comment by DaZombieKiller** [comment]
Hm, we might have to experiment with sending the pointer address over sockets or pipes instead then. I guess .NET Remoting as implemented by whatever version of Mono used in Unity 4/5 isn't compatible with .NET 4.x.

**Comment by spacehamster** [comment]
Also, i'm having trouble with Unity 4.7. It seems to crash when ever I try to get the next element from the TypeTreeIterator, and i'm not sure why.
https://github.com/spacehamster/TypeTreeDumper/blob/legacy/Unity/TypeTree.V1.cs#L105

**Comment by DaZombieKiller** [comment]
Does it work if you call `Next` at least once before calling `Children`?

**Comment by spacehamster** [comment]
No, it crashes on the first call to `Next` too. I don't think the order of the calls is the issue, because I was able to get it working with that order [here](https://github.com/spacehamster/NativeTypeTreeDumper/blob/master/NativeTypeTreeTools/NativeTypeTreeTools/TypeTree.cpp#L147), but something has gone wrong when i've translated it to C#.

**Comment by spacehamster** [comment]
I did some more debugging. On the managed side, the address of TypeTreeIterator i'm passing in is 0x28E60C, but the address that the function receives is 0x28E440. I can't figure out why the address is changing during the pinvoke.

![image](https://user-images.githubusercontent.com/3821056/94392568-ffe9c080-019b-11eb-8fdc-ba08d2e190a8.png)

![image](https://user-images.githubusercontent.com/3821056/94392694-4dfec400-019c-11eb-91da-21e3ded3bec3.png)

**Comment by spacehamster** [comment]
I wrote a test program
```c++
#define _CRT_SECURE_NO_WARNINGS
#include <stdio.h>
#include <stdlib.h>
#define EXPORT __declspec(dllexport) 

void WritePointer(const char* filePath, void* pointer) {
	FILE* file = fopen(filePath, "w");
	fprintf(file, "this=%p", pointer);
	fclose(file);
}

struct TypeTreeIterator {
	void* TypeTree;
	int TypeTreeList[4];
	EXPORT void TestThisCall1();
	EXPORT TypeTreeIterator TestThisCall2();
};


extern "C" {
	EXPORT void __thiscall TypeTreeIterator::TestThisCall1() {
		WritePointer("C:\\Files\\ThisCallTest1.txt", this);
	}

	EXPORT TypeTreeIterator __thiscall TypeTreeIterator::TestThisCall2() {
		WritePointer("C:\\Files\\ThisCallTest2.txt", this);
		TypeTreeIterator result{};
		return result;
	}

	EXPORT void __cdecl TestCdeclCall1(TypeTreeIterator* self) {
		WritePointer("C:\\Files\\ThisCdeclTest1.txt", self);
	}

	EXPORT TypeTreeIterator __cdecl TestCdeclCall2(TypeTreeIterator* self) {
		WritePointer("C:\\Files\\ThisCdeclTest2.txt", self);
		TypeTreeIterator result{};
		return result;
	}
}
```

and then in C#, I called it like
```csharp
[DllImport("TypeTreeHelper.dll", CallingConvention = CallingConvention.Cdecl)]
public static extern void TestCdeclCall1(ref TypeTreeIterator self);

[DllImport("TypeTreeHelper.dll", CallingConvention = CallingConvention.Cdecl)]
public static extern TypeTreeIterator TestCdeclCall2(ref TypeTreeIterator self);

[DllImport("TypeTreeHelper.dll",
	EntryPoint = "?TestThisCall1@TypeTreeIterator@@QAEXXZ",
	CallingConvention = CallingConvention.ThisCall)]
public static extern void TestThisCall1(ref TypeTreeIterator self);

[DllImport("TypeTreeHelper.dll",
	EntryPoint = "?TestThisCall2@TypeTreeIterator@@QAE?AU1@XZ",
	CallingConvention = CallingConvention.ThisCall)]
public static extern TypeTreeIterator TestThisCall2(ref TypeTreeIterator self);

void Test(ref TypeTreeIterator){
    TestCdeclCall1(ref it);
    TestCdeclCall2(ref it);
    TestThisCall1(ref it);
    TestThisCall2(ref it);
}
```
and the results are 
```
TestCdeclCall1=0028E60C
TestCdeclCall2=0028E60C
TestThisCall1=0028E60C
TestThisCall2=0028E41C
```

C# doesn't seem to be calling ThisCalls that return structs correctly. If thats the case, it would be possible to create a c++ library that exposes the TypeTreeIterator as a Cdecl, but i'd prefer to find a way to call the function directly.

**Comment by DaZombieKiller** [comment]
> C# doesn't seem to be calling ThisCalls that return structs correctly. If thats the case, it would be possible to create a c++ library that exposes the TypeTreeIterator as a Cdecl, but i'd prefer to find a way to call the function directly.

I think I actually remember seeing a bug report for this, but as far as I know it was only fixed in .NET Core. So for `__thiscall` methods returning structs larger than 8 bytes we might have to resort to native (or C++/CLR) after all.

**Comment by spacehamster** [comment]
I tested it with .NET core, and it works correctly. So I see two ways of doing it.
1. Convert the project to .NET core.

2. Create a helper library that wraps the function declaration of each method that needs to be called.

c++
```c++
ExampleStruct1 __cdecl WrappedFunction1(void* funcPtr, void* thisPtr){
	auto func = (ExampleStruct1 (__thiscall *)(void*, int))funcPtr;
	return func(thisPtr);
}
ExampleStruct2 __cdecl WrappedFunction1(void* funcPtr, void* thisPtr, int arg1){
	auto func = (ExampleStruct1 (__thiscall *)(void*, int, int))funcPtr;
	return func(thisPtr, arg1);
}
```
C#
```csharp
[DllImport("TypeTreeHelper32.dll", CallingConvention = CallingConvention.Cdecl)]
public static extern ExampleStruct1 WrappedFunction1(IntPtr funcPtr, IntPtr thisPtr);

[DllImport("TypeTreeHelper32.dll", CallingConvention = CallingConvention.Cdecl)]
public static extern ExampleStruct2 WrappedFunction2(IntPtr funcPtr, IntPtr thisPtr, int arg1);

void Invoke(){
	WrappedFunction1(resolver.Resolve("Function1"), &structVariable);
	WrappedFunction1(resolver.Resolve("Function2"), &structVariable, arg1);
}
```
Also, on the c++ side of things, its possible to create a general cdecl -> thiscall converter rather then wrapping each individual function, but its not pretty.
And you could probably use `System.Runtime.Emit` to partially apply the funcPtr, but that also wouldn't be pretty.

Moving to .NET core seems like the cleanest solution, is there any reason we wouldn't want to?

**Comment by DaZombieKiller** [comment]
Moving to .NET Core is definitely something I want to do, but .NET Core doesn't support remoting so we'll need to come up with a replacement for communicating between the processes.

**Comment by spacehamster** [comment]
Remoting (IPC) is just used for printing to the terminal at the moment, right? Could try to create a new terminal and print to that, or write to a log file and have the launcher read from that. Is there any need for IPC beyond that?

**Comment by DaZombieKiller** [comment]
It's also used to send across the output directory, but that can easily just be passed via the command line. Maybe it's worth playing around with `AttachConsole` to try and route output to the launcher's console directly? Even if that doesn't work though, there's always `AllocConsole` like you just suggested.

**Comment by DaZombieKiller** [comment]
As of https://github.com/DaZombieKiller/TypeTreeDumper/commit/04eea2285e8e249508e4052de4f4995f8c2bd99f I've now removed IPC-related code in preparation for a port to .NET Core (and perhaps CoreHook).

**Comment by DaZombieKiller** [comment]
I've just created a new `detours` branch which partially replaces EasyHook with MS Detours, which might allow hooking to work with Unity 3.4 and 3.5 (I haven't tested that yet) https://github.com/DaZombieKiller/TypeTreeDumper/tree/detours

**Comment by spacehamster** [comment]
I can't get MS Detours to build, is it missing any files from the repo?
the build script throws the error 
```
NMAKE : fatal error U1064: MAKEFILE not found and no target specified
Stop.
The system cannot find the path specified.
Microsoft (R) C/C++ Optimizing Compiler Version 19.27.29112 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

detours.c
c1: fatal error C1083: Cannot open source file: '..\..\detours.c': No such file or directory
```

**Comment by spacehamster** [comment]
The console sometimes throws this exception
```
System.IO.IOException: The request is not supported.

   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.__ConsoleStream.Write(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder)
   at System.IO.StreamWriter.Write(Char[] buffer, Int32 index, Int32 count)
   at System.IO.TextWriter.SyncTextWriter.WriteLine(String value)
   at System.Console.WriteLine(String value)
```

**Comment by DaZombieKiller** [comment]
> The console seems to have trouble handling multithreading

It might be worthwhile to make a wrapper `TextWriter` that directs all printing to a single worker thread then, if that's the case.

**Comment by spacehamster** [comment]
I figured it was a threading issue because the crash seemed to happen after switching threading contexts, but i've narrowed it down a bit.
```csharp
[SuppressMessage("Style", "IDE0060", Justification = "Required by EasyHook")]
public EntryPoint(RemoteHooking.IContext context, EntryPointArgs args)
{
    try
    {
        Kernel32.AttachConsole(Kernel32.ATTACH_PARENT_PROCESS);
        Console.WriteLine("Test1");
        Kernel32.FreeConsole();
        Kernel32.AttachConsole(Kernel32.ATTACH_PARENT_PROCESS);
        Console.WriteLine("Test2");
        OutputPath = args.OutputPath;
        module = Process.GetCurrentProcess().MainModule;
        resolver = new DiaSymbolResolver(module);
    }
    catch (Exception ex)
    {
        Console.Error.WriteLine(ex);
        throw;
    }
}
```
This throws the exception on `Console.WriteLine("Test2");`, but logs the exception to the console with `Console.Error.WriteLine(ex);`. I don't get why does stdout throw an exception, but its still able to write to stderr?

**Comment by spacehamster** [comment]
Changing `AttachToParentConsole` fixes the issue completely
```csharp
static void AttachToParentConsole()
{
    Kernel32.FreeConsole();
    Kernel32.AttachConsole(Kernel32.ATTACH_PARENT_PROCESS);
    TextWriter writer = new StreamWriter(Console.OpenStandardOutput()) { AutoFlush = true };
    Console.SetOut(writer);
}
```

**Comment by spacehamster** [comment]
MS Detours seems to be having trouble hooking 3.4-3.5 as well (it works fine on 2020.2.0b2.
When hooking 3.5, these exceptions appers in the unity logs (running multiple times produces different exceptions)
```
(0x00404583) c:\buildagent\work\b0bcff80449a48aa\runtime\allocator\memorymanager.cpp (618): MemoryManager::Allocate
(0x00402727) c:\buildagent\work\b0bcff80449a48aa\runtime\allocator\memorymanager.cpp (150): malloc_internal
(0x00716214) c:\program files (x86)\microsoft visual studio 9.0\vc\include\xstring (2093): std::basic_string<char,std::char_traits<char>,stl_allocator<char,39,4> >::_Copy
(0x00716468) c:\program files (x86)\microsoft visual studio 9.0\vc\include\xstring (1074): std::basic_string<char,std::char_traits<char>,stl_allocator<char,39,4> >::assign
(0x00B8F018) c:\buildagent\work\b0bcff80449a48aa\editor\src\application.cpp (691): Application::AfterEverythingLoaded
(0x0B3CCAA5) ((module-name not available)): (filename not available): (function-name not available)
(0x01AE95D2) (Unity): (filename not available): g_auedo
```
```
(0x00B8EF91) c:\buildagent\work\b0bcff80449a48aa\editor\src\application.cpp (679): Application::AfterEverythingLoaded
(0x0B4BCAA5) ((module-name not available)): (filename not available): (function-name not available)
(0x01AE95D2) (Unity): (filename not available): g_auedo
```
and our C# `AfterEverythingLoaded` function is never called.

**Comment by DaZombieKiller** [comment]
> 
> 
> Changing `AttachToParentConsole` to this fixes the issue completely, but I still don't really understand the issue.
> 
> `snip`

Interesting. I guess `AttachToConsole` invalidates the handle being used by the built in `Console.Out`, although it's strange that `Console.Error` continues to work. I'll add this fix in a bit.

> 
> 
> MS Detours seems to be having trouble hooking 3.4-3.5 as well (it works fine on 2020.2.0b2).
> When hooking 3.5, these exceptions appers in the unity logs (running multiple times produces different exceptions)
> 
> `snip`
> 
> `snip`
> 
> and our C# `AfterEverythingLoaded` function is never called.

That gets us further than with EasyHook at least, but it crashing like that is strange. Unless maybe Detours happens to hook in the same way EasyHook does, but without checking for this case?

**Comment by spacehamster** [comment]
AfterEverythingLoaded is `__thiscall` for versions 3.4-4.7 and `__cdecl` for versions 5.0-2020.2.0

**Comment by spacehamster** [comment]
I just noticed that CoreHook seems to use MS Detours as a foundation for it's native hooking, so I suspect that CoreHook will have similar results, but i'd like to confirm.

**Comment by spacehamster** [comment]
CoreHook's nuget packages seem out of date, which seems like it'll make using it a bit of a pain.
I found this project https://github.com/Reloaded-Project/Reloaded.Hooks , which looks promising. It also was unable to hook unity 3.5, but it looks like it would allow calling `__thiscalls` correctly from .NET Framework. I think the only downside to it is that it has some heavy-ish dependencies like FASM, but its probably not a huge issue.

**Comment by DaZombieKiller** [comment]
`Reloaded.Hooks` definitely looks like it'll be more useful overall, it offers a LOT more control. When you tried to hook Unity 3.5, which result did you get? An error about being unable to hook or just incorrect runtime behaviour?

**Comment by spacehamster** [comment]
The error I get with Reloaded.Hooks is the same one I get with MS Detours, a crash in the unity log
```
========== OUTPUTING STACK TRACE ==================

(0x00B8EF9C) c:\buildagent\work\b0bcff80449a48aa\editor\src\application.cpp (679): Application::AfterEverythingLoaded
(0x0EEECAA5) ((module-name not available)): (filename not available): (function-name not available)
(0x01AE95D2) (Unity): (filename not available): g_auedo

========== END OF STACKTRACE ===========
```
and then the app closes.

[Log.txt](https://github.com/DaZombieKiller/TypeTreeDumper/files/5337435/Log.txt)

The crash log says 
```
Unity.exe caused an Access Violation (0xc0000005)
  in module Unity.exe at 0023:00b8ef90.
```
(00b8ef90 is the entry point of the original function AfterEverythingLoaded)

[error.log](https://github.com/DaZombieKiller/TypeTreeDumper/files/5337427/error.log)

If I attach a debugger for either MS Detours or Reloaded.Hooks, no log file or crash report is created and the breakpoint at AfterEverythingLoaded is never hit.

**Comment by DaZombieKiller** [comment]
The strange part is that this seems to apply to _all_ functions in Unity 3.x, I can't seem to hook anything without a crash.

**Comment by spacehamster** [comment]
I'm wondering if perhaps there is some sort of memory protection at play, like writing the hook triggers some protection that marks the code non-executable and then triggers an access violation when it tries to execute it?

**Comment by DaZombieKiller** [comment]
After messing with it for a while, I'm definitely starting to think that could be the case. Do you think we could fix it using `VirtualProtect`?

**Comment by DaZombieKiller** [comment]
There's definitely something up with the editor executable, but hooking the player executable works. I've just pushed a few changes to the `detours` branch for that. It reaches the dumping stage but since support for the 3.x type tree isn't implemented, it obviously fails once it tries to generate one for `GameObject`.

Here's the dummy 3.5.7 built project I used for testing (it's a `.7z`, I renamed it to `.zip` so GitHub would let me upload it)
[Unity357_DummyApp.zip](https://github.com/DaZombieKiller/TypeTreeDumper/files/5347211/Unity357_DummyApp.zip)

**Comment by spacehamster** [comment]
> There's definitely something up with the editor executable, but hooking the player executable works

Comparing the disassembly of the editor and the player, theres a huge difference, there could be some compiler obfuscation going on.
Editor
![image](https://user-images.githubusercontent.com/3821056/95641844-65b93f00-0af0-11eb-8cd7-45ace780efc6.png)
Player
![image](https://user-images.githubusercontent.com/3821056/95641855-7964a580-0af0-11eb-85a4-4b46f5a0db2d.png)

**Comment by spacehamster** [comment]
I tried calling [VirtualQuery](https://docs.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-virtualquery) on the function address. 
In 3.5
```
VirtualQuery. Result=True. Address=0xB8EF90. Module=0x400000
BaseAddress: 12115968 (0xB8E000)
AllocationBase: 4194304 (0x400000)
AllocationProtect: 128 (0x80)
PartitionId: 40960 (0xA000)
RegionSize: 4096 (0x1000)
State: 32 (0x20)
Protect: 16777216 (0x1000000)
Type: 0 (0x0)
```
and in 4.7.0
```
VirtualQuery. Result=True. Address=0x10149A0. Module=0x400000
BaseAddress: 16859136 (0x1014000)
AllocationBase: 4194304 (0x400000)
AllocationProtect: 128 (0x80)
PartitionId: 36864 (0x9000)
RegionSize: 4096 (0x1000)
State: 32 (0x20)
Protect: 16777216 (0x1000000)
Type: 0 (0x0)
```

Calling VirtualProtect with `flNewProtect=0x80`, returns `lpflOldProtect=0x20` and changes the State field to 0x80, but does not resolve the issue. Because VirtualQuery has the same values for 3.5 and 4.7, i'm not convinced the issue is virtual memory protection.

0x20 = PAGE_EXECUTE_READ
0x80 = PAGE_EXECUTE_WRITECOPY

**Comment by DaZombieKiller** [comment]
Yeah, I did a bit of experimentation with changing the virtual protection flags and had the same results. So I guess it must be some form of obfuscation after all. Since we can still dump type trees from the player I'm probably not going to bother investigating further, at least not for now.

**Comment by spacehamster** [comment]
I believe we can still dump types from the editor using the script loader entry point method too, we just can't hook any editor functions.

**Comment by spacehamster** [comment]
To add 3.x/4.x dumping support, we either need to move to Reloaded.Hooks or move to .net core + CoreHook. Since CoreHook's nuget packages are out of date, my preference is .net framework + Reloaded.Hooks.

**Comment by DaZombieKiller** [comment]
Yeah, CoreHook's main package is up to date I believe, but all of its other NuGet packages are super out of date which makes it almost impossible to use in this project without using a mess of submodules. Reloaded.Hooks is definitely the one that I think we should move to.

**Comment by DaZombieKiller** [comment]
> I believe we can still dump types from the editor using the script loader entry point method too, we just can't hook any editor functions.

Right, that's also an option. We wouldn't be able to dump abstract types since that requires hooking, but that's not necessary for existing tools anyway (and there's always the option of just using the player for that, because editor-only types aren't needed at all.)

**Comment by spacehamster** [comment]
I wanted to use some of the code in other projects, is it possible to add a license?

**Comment by DaZombieKiller** [comment]
Sure, I'll add the MIT license to the repo

**Comment by DaZombieKiller** [comment]
Done https://github.com/DaZombieKiller/TypeTreeDumper/commit/a7657a1de2ed023e441b2fa4a54ad4d46719ae86

**Comment by DaZombieKiller** [comment]
Here's a JSON file that maps Unity version and build numbers to directory names under `Unity_x64_mono.pdb` on the symbol server: https://gist.github.com/DaZombieKiller/0a848dc59436ce23ef86e4206dfa8979

Interestingly, this reveals that it only contains PDBs for Unity 2018.1 through to Unity 2019.1. Generating these JSON files is a long process, so I'm still working on generating one for `UnityPlayer_Win64_development_mono_x64.pdb`.

**Comment by spacehamster** [comment]
> Interestingly, this reveals that it only contains PDBs for Unity 2018.1 through to Unity 2019.1. 
That matches up with what I experienced.

I'm looking at adding dumping support for versions older then 3.4-4.7, but i'm having a bit of trouble. The TypeTree format has changed so there are no TypeTreeNodes any more. The main issue is that TypeTree children are stored in an std::list, but it seems that all of the std::list functions are inlined, so there are no suitable functions to call. I had planned to use a TypeTreeIterator to access children instead, but it appears that was only introduced in 4.7, so the 3.4-4.6 would still be unsolved.

![image](https://user-images.githubusercontent.com/3821056/97159656-4cdaaa00-17cf-11eb-991b-ca7429b33a2e.png)

I suspect even if you used c++ to access it with the std::list functions like this does https://gist.github.com/robert-nix/7db0145e809b692b63f2#file-unitystructgen-cpp-L72, it would not work unless you used the same version of the standard library runtime that unity compiled against.

**Comment by DaZombieKiller** [comment]
Yeah, I've been working on a refactoring of some of the engine interop code and noticed the usage of `std::list` in older versions. I've pretty much hit the same wall there. I guess it could be possible to make a native DLL that exposes an API for interacting with `std::list`, and then we just need to ensure that it compiles under the same version of the standard library.

**Comment by spacehamster** [comment]
Another idea I had considered was to try and reimplment the std::list standard library functions in C#, but I did not have much success when I tried that idea for std::string.c_str, so i'm not especially optimistic about that approach.

**Comment by DaZombieKiller** [comment]
I considered doing that, but I think for now it's very much a last resort approach. It'd be nice to have a managed implementation at some point, but I'd like to exhaust the simpler options first.

**Comment by spacehamster** [comment]
There is a function `void __cdecl WalkTypeTree(class TypeTree const &,unsigned char const *,int *))` which seems like it might help, but I can't figure out what it is actually doing.

**Comment by DaZombieKiller** [comment]
Since the API exposed by the generated Dia2Lib is messy and doesn't expose some things properly, I've been working on a replacement for it: https://github.com/DaZombieKiller/Dia2. It has XML documentation comments for the entire public API surface based on the [DIA SDK Reference](https://docs.microsoft.com/en-us/visualstudio/debugger/debug-interface-access/debug-interface-access-sdk-reference).

I haven't finished porting everything across though, so it still depends on Dia2Lib for now. I'm currently using this in a GUI program for viewing the contents of PDB files (and for generating C++ and C# structures from them) which I'll also put up on GitHub once it's ready.

Ideally we should be able to turn the `Symbol` class into a bunch of other classes (`CompilandSymbol`, `PublicSymbol`, etc) to make it clearer which fields are actually usable on a particular symbol.

I also noticed that Ghidra has support for importing `.pdb.xml` files which it generates from `.pdb` files on Windows. I'm going to see if I can allow my tool to output in this format too with extra customization, so we can merge type information.

**Comment by spacehamster** [comment]
I've been playing around with the older versions of unity, and have had some success accessing TypeTreeLists like this:
```cpp
#define _ITERATOR_DEBUG_LEVEL 0
#include <string>
#include <list>
#define EXPORT  extern "C" __declspec(dllexport)

struct TypeTree;

template <class T = char>
class stl_allocator : public std::allocator<T>
{
public:
	template<class _Other>
	struct rebind
	{
		typedef stl_allocator_new<_Other> other;
	};

	stl_allocator_new()
	{
	}
	stl_allocator_new(const std::allocator<T>&)
	{
	}
};

typedef std::list<TypeTree, stl_allocator<TypeTree>> TypeTreeList;
typedef std::basic_string<char, std::char_traits<char>, stl_allocator<char>> TypeTreeString;

struct TypeTree
{
	TypeTreeList m_Children;
	int _dummy1;
	int _dummy2;
	TypeTree* m_Father;
	TypeTreeString m_Type;
	int _dummy3;
	TypeTreeString m_Name;
	int _dummy4;
	int m_ByteSize;
	int m_Index;
	int m_IsArray;
	int m_Version;
	int m_MetaFlag;
	int m_ByteOffset;
	void* m_DirectPtr;
};


EXPORT const char* StringToCStr(std::string* str) {
	return str->c_str();
}

EXPORT unsigned int ListSize(TypeTreeList* list) {
	return list->size();
}

EXPORT TypeTreeNew& ListGet(TypeTreeList* list, int index) {
	int i = 0;
	for (auto it = list->begin(); it != list->end(); it++)
	{
		if (i == index) {
			return *it;
		}
		i++;
	}
	TypeTreeNew result;
	return result;
}
```
csharp
```csharp
[DllImport("StlHelper", CallingConvention = CallingConvention.Cdecl)]
internal static extern IntPtr StringToCStr(void* str);

[DllImport("StlHelper", CallingConvention = CallingConvention.Cdecl)]
internal static extern int ListSize(void* list);

[DllImport("StlHelper", CallingConvention = CallingConvention.Cdecl)]
internal static extern TypeTree* ListGet(void* list, int index);
```
It doesn't work when i add `int rootref;` to the stl_allocator, it causes the struct alignment to become invalid, i'm not quite sure why.

TypeTreeList seems to be implemented as a doubly linked list, and i think it would be cleaner to access it purely from c#, like this
```csharp
unsafe struct TypeTreeList
{
	public TypeTreeNode* Head;
	public unsigned int Size;
	public int Padding1;
	public int Padding2;
}
unsafe struct TypeTreeNode
{
	public TypeTreeNode* Next;
	public TypeTreeNode* Prev;
	public TypeTree* Value;
}
```

**Comment by spacehamster** [comment]
Did some more testing, and was able to successfully walk the tree without the c++ helpers like this
```csharp
public void CreateNodes(ManagedTypeTree owner, ref List<TypeTreeNode> nodes, ref TypeTree tree, int level = 0)
{
	var typeIndex = GetOrCreateStringIndex(tree.m_Type);
	var nameIndex = GetOrCreateStringIndex(tree.m_Name);
	var nodeImpl = new TypeTreeNode.V1(
		version: (short)tree.m_Version,
		level: (byte)level,
		typeFlags: (TypeFlags)tree.m_IsArray,
		typeStrOffset: typeIndex,
		nameStrOffset: nameIndex,
		byteSize: tree.m_ByteSize,
		index: tree.m_Index,
		metaFlag: tree.m_MetaFlag);
	nodes.Add(new TypeTreeNode(nodeImpl, owner));

	TypeTreeListNode node = tree.m_Children.Head;
	for(int i = 0; i < tree.m_Children.Size; i++)
	{
		node = node->Next;
		TypeTree child = node->Value;
		CreateNodes(owner, ref nodes, ref child, level + 1);
	}
}
```
I think accessing it from c# is the cleanest solution

**Comment by DaZombieKiller** [comment]
Nice! So is this for Unity 3.4? Or is there an even earlier version available on Windows?

> I think accessing it from c# is the cleanest solution

I agree, it also keeps the build system simple since we don't need to include a C++ project.

**Comment by spacehamster** [comment]
I've only tested this in 4.5, but i'm hoping the TypeTreeList layout will be the same with 3.4 and the same method will work there too.

**Comment by DaZombieKiller** [comment]
Here's part of the code for the 3.4 TypeTree from the refactor I've been working on:
```cs
using System;

namespace Unity.V3_4
{
    public unsafe partial struct TypeTree
    {
        public CppList Children;
        public TypeTree* Father;
        public CppString Type;
        public CppString Name;
        public int ByteSize;
        public int Index;
        public int IsArray;
        public int Version;
        public TransferMetaFlags MetaFlag;
        public int ByteOffset;
        public IntPtr DirectPtr;
    }
}
```
Does that look similar?

**Comment by spacehamster** [comment]
yes
```csharp
internal struct TypeTreeString
{
	fixed byte Data[28];
};

internal unsafe struct TypeTreeList
{
	public TypeTreeListNode* Head;
	public uint Size;
	public int Padding1;
	public int Padding2;
};

internal unsafe struct TypeTreeListNode
{
	public TypeTreeListNode* Next;
	public TypeTreeListNode* Prev;
	public TypeTree Value;
}

internal struct TypeTree
{
	public TypeTreeList m_Children;
	public TypeTree* m_Father;
	public TypeTreeString m_Type;
	public TypeTreeString m_Name;
	public int m_ByteSize;
	public int m_Index;
	public int m_IsArray;
	public int m_Version;
	public TransferMetaFlags m_MetaFlag;
	public int m_ByteOffset;
	public void* m_DirectPtr;
}```

**Comment by spacehamster** [comment]
Also TypeTreeString's implementation seems to be identical to this https://stackoverflow.com/questions/40393350/why-does-microsofts-implementation-of-stdstring-require-40-bytes-on-the-stack
so if there is ever a version that is missing c_str i think it should be possible to access it from C#.

**Comment by spacehamster** [comment]
How are you running TypeTreeDumper on the 3.4 player? Are you building a custom .exe or can you run `Data\PlaybackEngines\windowsdevelopmentstandaloneplayer\player_win.exe` directly?

**Comment by DaZombieKiller** [comment]
> 
> 
> How are you running TypeTreeDumper on the 3.4 player? Are you building a custom .exe or can you run `Data\PlaybackEngines\windowsdevelopmentstandaloneplayer\player_win.exe` directly?

I just run it on `player_win.exe` and hook `PlayerInitEngineNoGraphics` instead of `AfterEverythingLoaded`.

**Comment by spacehamster** [comment]
I get this message trying to run player_win.exe
![image](https://user-images.githubusercontent.com/3821056/97849966-7bb0cd00-1d47-11eb-8c67-ec6b63934279.png)
I assume i'm meant to build a game then?

**Comment by DaZombieKiller** [comment]
Ah, yeah. You need to build a small dummy game so that the player will boot.

**Comment by spacehamster** [comment]
So I got a bit suspicious when the disassembly in visual studio didn't match up with the disassembly in ghidra in Unity 3.5. If you dump the raw byte for the function `AfterEverythingLoaded` before `RemoteHooking.WakeUpProcess` is called, you get the same results as ghidra.
![image](https://user-images.githubusercontent.com/3821056/98498605-4c481600-229b-11eb-8e27-eae3e6782903.png)
Raw Memory
```
AfterEverythingLoaded: 00B8EF90
0000: 427D7493AD2BF468
0008: 66F64A989697643D
0010: FD51BE92DC7BD246
0018: 9F49EE2112D0B664
0020: 2E9C4445B69B03FB
0028: EE03BD39E50191F9
0030: F464B72DCF81A5BD
0038: A8BB68AF8A960BB8
0040: C10DC66900C5867D
0048: E2E953E458CF3BD4
0050: 5D013CF256DB43C6
0058: 4BEC6066AC4AABC3
0060: CB906C5650FB92F6
0068: 3D2CF452767CB545
0070: 9E0EC533BB7C9DD2
0078: E6FE955DFDFC8A09
```
Putting those bytes through an online decompiler says its junk.
But if you wait until after `-executeMethod Loader.Load` is called with the legacy loading method, the raw memory looks like this
```
AfterEverythingLoaded: 00B8EF90
0000: 558BEC6AFF68F8ED
0008: 160164A100000000
0010: 5083EC5CA1080B6B
0018: 0133C58945F05356
0020: 57508D45F464A300
0028: 000000E890E7CCFF
0030: 8BC8E819E8CCFFE8
0038: 042760008B702433
0040: DBBF0F000000C745
0048: D42B000000897DEC
0050: 895DE8885DD8895D
0058: FCE8727DE5FF8B48
0060: 108B50140FACD102
0068: C1EA0280E1018D4D
0070: D474096A0468781E
0078: 3E01EB076A076870
```
and running that through a disassembler produces actual asm code.

```assembly
0:  55                      push   ebp
1:  8b ec                   mov    ebp,esp
3:  6a ff                   push   0xffffffff
5:  68 f8 ed 16 01          push   0x116edf8
a:  64 a1 00 00 00 00       mov    eax,fs:0x0
10: 50                      push   eax
11: 83 ec 5c                sub    esp,0x5c
14: a1 08 0b 6b 01          mov    eax,ds:0x16b0b08
19: 33 c5                   xor    eax,ebp
1b: 89 45 f0                mov    DWORD PTR [ebp-0x10],eax
1e: 53                      push   ebx
1f: 56                      push   esi
20: 57                      push   edi
21: 50                      push   eax
22: 8d 45 f4                lea    eax,[ebp-0xc]
25: 64 a3 00 00 00 00       mov    fs:0x0,eax
2b: e8 90 e7 cc ff          call   0xffcce7c0
30: 8b c8                   mov    ecx,eax
32: e8 19 e8 cc ff          call   0xffcce850
37: e8 04 27 60 00          call   0x602740
3c: 8b 70 24                mov    esi,DWORD PTR [eax+0x24]
3f: 33 db                   xor    ebx,ebx
41: bf 0f 00 00 00          mov    edi,0xf
46: c7 45 d4 2b 00 00 00    mov    DWORD PTR [ebp-0x2c],0x2b
4d: 89 7d ec                mov    DWORD PTR [ebp-0x14],edi
50: 89 5d e8                mov    DWORD PTR [ebp-0x18],ebx
53: 88 5d d8                mov    BYTE PTR [ebp-0x28],bl
56: 89 5d fc                mov    DWORD PTR [ebp-0x4],ebx
59: e8 72 7d e5 ff          call   0xffe57dd0
5e: 8b 48 10                mov    ecx,DWORD PTR [eax+0x10]
61: 8b 50 14                mov    edx,DWORD PTR [eax+0x14]
64: 0f ac d1 02             shrd   ecx,edx,0x2
68: c1 ea 02                shr    edx,0x2
6b: 80 e1 01                and    cl,0x1
6e: 8d 4d d4                lea    ecx,[ebp-0x2c]
71: 74 09                   je     0x7c
73: 6a 04                   push   0x4
75: 68 78 1e 3e 01          push   0x13e1e78
7a: eb 07                   jmp    0x83
7c: 6a 07                   push   0x7
7e: 68                      .byte 0x68
7f: 70                      .byte 0x70
```

I suspect the exe is encrypted, compressed or somehow obfuscated and the hooks can't be set until after the valid executable has been loaded into memory. It makes reverse engineering difficult because the exe on disk is not valid.

**Comment by spacehamster** [comment]
I got a proof of concept for hooking to work in 3.5 like this
```csharp
[SuppressMessage("Style", "IDE0060", Justification = "Required by EasyHook")]
public void Run(RemoteHooking.IContext context, EntryPointArgs args){
	...snip...
	OnEngineInitialized += ExecuteDumper;
	RemoteHooking.WakeUpProcess();
	if (VersionInfo.FileMajorPart == 3)
	{
		HookAfterEverythingLoadLegacy();
	}
	Thread.Sleep(Timeout.Infinite);
	...snip...
}
unsafe void HookAfterEverythingLoadLegacy()
{
	Console.WriteLine("Trying to hook AfterEverythingLoaded");
	resolver.TryResolveFirstMatching(new Regex(Regex.Escape("?AfterEverythingLoaded@Application@") + "*"), out var address);
	byte* buffer = (byte*)address;
	byte[] matchingCodes = new byte[] {
		0x55, 0x8B, 0xEC, 0x6A, 0xFF, 0x68, 0xF8, 0xED
	};
	//Wait for target to be decrypted
	bool validCodes = false;
	while (!validCodes)
	{
		bool foundMatch = true;
		for(int i = 0; i < matchingCodes.Length; i++)
		{
			if (buffer[i] != matchingCodes[i]) foundMatch = false;
		}
		validCodes = foundMatch;
	}
	Console.WriteLine("Hooking AfterEverythingLoaded");
	AfterEverythingLoadedHook = LocalHook.Create(address, new AfterEverythingLoadedDelegate(AfterEverythingLoaded), null);
	AfterEverythingLoadedHook.ThreadACL.SetExclusiveACL(Array.Empty<int>());
	Console.WriteLine("Hooked AfterEverythingLoaded");
}
```

**Comment by spacehamster** [comment]
I'm wondering which approach you'd prefer, dumping the 3.5 editor or dumping the 3.5 player?

**Comment by DaZombieKiller** [comment]
Dumping from the editor is more consistent, so in general I think it's nice to support it. Ideally we can support both, so I think it's a good idea to implement your solution.

**Comment by spacehamster** [comment]
For initializing the AfterEverythingLoadCallback, do you prefer the commandline `-executeMethod` method or the wait for decrypt and then easyhook method?
The commandline method seems safer as the wait for decrypt method introduces a race condition were it is possible for the method to be decrypted and then the hook be set *after* the target function is called.

**Comment by DaZombieKiller** [comment]
Yeah, the `executeMethod` approach seems better for now and far less error prone. We'll just need to move all of the hooking stuff out into its own function so AfterEverythingLoaded can call it on 3.5.

**Comment by DaZombieKiller** [comment]
I did a bit of thinking and figured out that we can pass data from the dumper to the editor by using process environment variables. I've replaced the template string approach with this in https://github.com/DaZombieKiller/TypeTreeDumper/commit/ddec84f0b99c1238ab1a41ad72b0b82bd3b10e5c.

**Comment by spacehamster** [comment]
> Interesting, so 3.4 passes around allocator pointers instead then? Hopefully there's an easy way to access the right one. If there's no global function for retrieving it, we could hook a function that is always called in order to grab whatever allocator is passed to it.

If the base allocator passed into Object::Produce is null, it uses a default allocator. I'm gonna try that and see  if it works
![image](https://user-images.githubusercontent.com/3821056/98648305-37967b80-238a-11eb-8d44-d8b1682fe76f.png)

**Comment by DaZombieKiller** [comment]
Just wanted to mention that this experimental build of Unity 2020.2 contains a completely unstripped PDB: https://forum.unity.com/threads/experimental-contacts-modification-api.924809/#post-6427604
